### ISSUE TITLE
Should we have instances only included in references when they are either the top level module or a child of a definition

### ISSUE THREAD

**Issue by andrewmkeller** [byuccl/spydrnet#114]

**Comment by thunder-hammer** [comment]
Good question! I think our current implementation has all instances whether they are used or not included in the definition's references.

Do you have a specific use case in mind?

Advantages I can think of would be that unused instances would be ignored when iterating through references. For example in cases where the instances are not actually used uniquify would have to uniquify less. Counting instances of the definition would also be simplified. And after all if it's not actually used then why should the definition keep track of it.

On the other hand this might also introduce a little more complexity to the end user, For example if the current add_port algorithm iterates on the references list. it might be more confusing if some things that used to be instances of the definition no longer mirrored it because they were not children at the time.

If an algorithm want to operate on just instantiated instances as in uniquify. it could look at the parent pointer of the instance to see if the instance is used. this would let users who want it to iterate through everything to do it, and users who need it to only work on those that are instanced could check as they iterate

Considering this I think that we should have the definitions' references include all the instances whether they are children or top or not.

TLDR:
That's a good thing to keep in mind and there are advantages on both sides, but I think having the definition's references include even unused instances gives the most flexibility and consistency to the end users.
